### PR TITLE
Add tests in /examples/restricted-assets

### PR DESCRIPTION
**Pull request by ratik21**

### CODE REVIEW

File: examples/restricted-assets/assets/poi-approval.teal
Comment (amityadav0) [inline examples/restricted-assets/assets/poi-approval.teal:231]:
```suggestion
// Note: the receiver of the asset is asset xfer
```

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:38]: how about renaming this to `beforeEach` ? and remove `beforeAll`.

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:87]: shouldn't we add opt-in for alice as well ?

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [inline packages/runtime/src/runtime.ts:360]: can you please elaborate why are we doing this ?

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (amityadav0) [inline packages/runtime/test/src/interpreter/opcode-list.ts:1982]: why are doing this ?
i think `const op = new Txna(["Accounts", "1"], 1, interpreter);` means second account.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:339]: yes, confirmed.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:360]: Because for creator even if the asset is default frozen, for his holding the frozen is `false`. Confirmed by testing.
```
Note: A frozen account can always close out to the asset creator
```
Reference: [https://developer.algorand.org/tutorials/asa-javascript/](https://developer.algorand.org/tutorials/asa-javascript/)

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (ratik21) [reply examples/restricted-assets/test/asset-txfer-test.js:130]: To pay tx fees. Initially it has 0 balance.

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (ratik21) [reply examples/restricted-assets/test/asset-txfer-test.js:261]: ok

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (ratik21) [reply packages/runtime/test/src/interpreter/opcode-list.ts:1982]: I though so too. But `Txna Account 0` means sender's address, and the above means first account. 
Tested (on blockchain) using: `gtxn 0 Accounts 1` (this pushed the first account from the account array). Something seems wrong. Will confirm on discord as well.

Comment (ratik21) [reply packages/runtime/test/src/interpreter/opcode-list.ts:1982]: Update: Confimed `const op = new Txna(["Accounts", "1"], 1, interpreter);` means first account.

Will do this in a new PR first, with more tests. We also need to update `Txn`, `Gtxn` as well.

Comment (amityadav0) [reply packages/runtime/test/src/interpreter/opcode-list.ts:1982]: i think we already have tests but they are written according to code.
or you can update on this PR as well, In interpreter file we already have `getAccount` function.
update lines with this.getAccount(1,line).your call

Comment (ratik21) [reply packages/runtime/test/src/interpreter/opcode-list.ts:1982]: Yes, but we need to add "more" tests with the correct behaviour, since we previously assumed wrong one, according to which the tests were written (your initial comment). 
Also found out that `Txn`, `Gtxn` needs to be updated as well because they also support array arguments. (eg. `txn Accounts 0`, `gtxn 0 Accounts 0` also works), so it's better to handle these fixes along with this fix in a new PR.

Comment (amityadav0) [reply packages/runtime/test/src/interpreter/opcode-list.ts:1982]: like i said, your call

File: examples/restricted-assets/assets/poi-approval.teal
Comment (amityadav0) [inline examples/restricted-assets/assets/poi-approval.teal:232]: we have support for array in gtxna. right ?

Comment (amityadav0) [reply examples/restricted-assets/assets/poi-approval.teal:232]: or we will update it after fix gets merged ?

Comment (ratik21) [reply examples/restricted-assets/assets/poi-approval.teal:232]: Yes, we already had support for array in `gtxna`. After fix we will have support for array with `txn`, `gtxn` as well. 
Updated TEAL code here because i think `gtxna 0 Accounts 1` is more intuitive than `gtxn 0 Accounts 1`. Both will work after the fix. Let me know what you think we should keep here

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:80]:
```suggestion
  // Update account state
```

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:62]:
```suggestion
```

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:153]:
```suggestion
    /* Updating clawback again should throw error now (since we set the asset manager to ``) */
```

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:189]: Please add 1000 assets revoked from clawback

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (amityadav0) [inline examples/restricted-assets/test/asset-txfer-test.js:97]: Please modify this line to revoke assets

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [inline packages/runtime/src/runtime.ts:822]: can you please explain this ?

Comment (amityadav0) [reply packages/runtime/src/runtime.ts:822]: should we set this.ctx.tx back to previous after we are done with lsig ?

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (ratik21) [reply examples/restricted-assets/test/asset-txfer-test.js:188]: Adding

File: examples/restricted-assets/test/asset-txfer-test.js
Comment (ratik21) [reply examples/restricted-assets/test/asset-txfer-test.js:96]: Will add revoking instead of modifying. From the top level it's just a transfer of asset. Revoking via clawback escrow is internal logic

Comment (amityadav0) [reply examples/restricted-assets/test/asset-txfer-test.js:96]: ok

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:822]: > can you please explain this ?

We need to set the current transaction (for correct `GroupIndex`) before executing stateless TEAL

> should we set this.ctx.tx back to previous after we are done with lsig ?

Not sure if setting previous config the right way. What if the first tx is the stateless one?

Comment (amityadav0) [reply packages/runtime/src/runtime.ts:822]: for the ssc `tx` is first transaction in `gtxs`, right ?

Comment (ratik21) [reply packages/runtime/src/runtime.ts:822]: > should we set this.ctx.tx back to previous after we are done with lsig ?

Doing this.

**Review by amityadav0 — APPROVED** [review]
Please verify that tests are working after updating builds.

**Pull request merged**
